learning/tour-of-beam/frontend/lib/repositories/client/client.dart

```diff
@@ -0,0 +1,26 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import '../../models/content_tree.dart';
+import '../models/get_sdks_response.dart';
+
+abstract class TobClient {
```

Comment:
I like that we are using an abstract class to encapsulate connection to the backend service.  However, is there any way we can follow the same client/service model we use with the current Beam playground?

Reply:
We do have the same entity caches and clients in Playground. The difference is that in Playground we additionally have a repository layer between an entity cache and the abstract network client. It is a useful abstract for an entity cache to not be aware if it gets objects from the network or from a local DB. However, network is the only implementation in Playground so far, and until we have two implementations it's better to not add abstractions. For example, in Playground we have `CodeRespository` which has `runCode` method. It should be an action and not a repository intent, but it was modeled after `ExampleRepository` to just have the same number of abstractions. If we need a layer between an entity cache and the network, I would rather wait for a hands-on with local storage before adding an extra layer here just to see what are the differences in requirements for the implementations.

So I would rather make Playground closer to the current model of ToB.

Reply:
I'm referring to either gRPC, currently used by Beam Playground or an Open API specification contract between the service and client.

Reply:
We rejected gRPC in favor of cloud functions for the backend, and it is in the TDD.

As for OpenAPI, we can technically generate Dart files with something like https://pub.dev/packages/openapi_generator
if we had the spec from the backend team.

@eantyshev is it possible to make an OpenAPI spec for these functions?

Reply:
IMHO OpenApi spec is good when supported with code-generated handlers and models.
We don't have that on backend side, but we provide a [README that exposes API samples](https://github.com/apache/beam/pull/23318/files?short_path=5e42102#diff-5e421024a73e17f28a1e7795af840bb51d733226805f18dbaaf9304bfae8b25c)

Reply:
AFAIK the native Go way to keep API models is like this: https://github.com/apache/beam/blob/master/learning/tour-of-beam/backend/internal/entity.go

Reply:
To summarize, we could have an OpenAPI spec to generate frontend files from it. But backend is built in another way, and its code cannot be generated from OpenAPI.

So we have 3 locations to maintain that need to agree on the API:

1. The backend code itself.
2. The README of the backend.
3. The code in the frontend.

If we introduce OpenAPI spec, the 3rd location will change:

1. The backend code itself.
2. The README of the backend.
3. The OpenAPI YAML spec.

But it's still three locations, the 1st location is not going anywhere.

Reply:
Thank you for thinking through this so carefully.